Source/hydro/slope.cpp

```diff
@@ -280,60 +227,120 @@ Castro::pslope(const Box& bx, const int idir,
     AMREX_PARALLEL_FOR_3D(bx, i, j, k,
     {
 
-      // First compute Fromm slopes
+      Real p0;
+      Real pp1;
+      Real pm1;
+      Real pp2;
+      Real pm2;
 
-      // Dp at i+1
+      Real p0_hse;
+      Real pp1_hse;
+      Real pp2_hse;
+      Real pm1_hse;
+      Real pm2_hse;
 
-      // we need dp_{i+1} = p_{i+1} - p_i
-      // and     dp_{i+2} = p_{i+2} - p_{i-1}
-      //
-      // then we compute
-      //   Dp_{i+1} = min{ 2 |dp_{i+1}|, 2 |dp_{i+2}|, 1/2 |dp_{i+1} + dp_{i+2}|}
-      //
-      // for this construction, we take p_{i+1} as the reference
-      // and subtract off the HSE state integrated from this from each zone center
-      //
+      if (idir == 0) {
 
-      Real dlftp1;
-      Real drgtp1;
+        p0_hse = q_arr(i,j,k,QPRES);
 
-      if (idir == 0) {
+        pp1_hse = p0_hse + 0.25_rt*dx[0] *
+          (q_arr(i,j,k,QRHO) + q_arr(i+1,j,k,QRHO)) *
+          (src(i,j,k,QU) + src(i+1,j,k,QU));
+        pp2_hse = pp1_hse + 0.25_rt*dx[0] *
+          (q_arr(i+1,j,k,QRHO) + q_arr(i+2,j,k,QRHO)) *
+          (src(i+1,j,k,QU) + src(i+2,j,k,QU));
+
+        pm1_hse = p0_hse - 0.25_rt*dx[0] *
+          (q_arr(i,j,k,QRHO) + q_arr(i-1,j,k,QRHO)) *
+          (src(i,j,k,QU) + src(i-1,j,k,QU));
+        pm2_hse = pm1_hse - 0.25_rt*dx[0] *
+          (q_arr(i-1,j,k,QRHO) + q_arr(i-2,j,k,QRHO)) *
+          (src(i-1,j,k,QU) + src(i-2,j,k,QU));
 
-        // this is dp_{i+1} = p_{i+1} - p_i, but we subtract off the
-        // HSE state integrated from p_{i+1}
-        dlftp1 = q_arr(i+1,j,k,QPRES) - q_arr(i,j,k,QPRES);
-        dlftp1 += -0.25_rt * (q_arr(i+1,j,k,QRHO) + q_arr(i,j,k,QRHO)) *
-          (src(i+1,j,k,QU)+src(i,j,k,QU))*dx[0];
+        p0 = 0.0_rt;
+        pp1 = q_arr(i+1,j,k,QPRES) - pp1_hse;
+        pp2 = q_arr(i+2,j,k,QPRES) - pp2_hse;
 
-        // this is dp{i+2} = p_{i+2} - p_{i+1}, but we subtract off the
-        // HSE state integrated from p_{i+1}
-        drgtp1 = q_arr(i+2,j,k,QPRES) - q_arr(i+1,j,k,QPRES);
-        drgtp1 += -0.25_rt * (q_arr(i+1,j,k,QRHO) + q_arr(i+2,j,k,QRHO)) *
-          (src(i+1,j,k,QU)+src(i+2,j,k,QU))*dx[0];
+        pm1 = q_arr(i-1,j,k,QPRES) - pm1_hse;
+        pm2 = q_arr(i-2,j,k,QPRES) - pm2_hse;
+
+        if (i == domlo[0] && lo_bc_test) {
```

Comment:
If we enforce the symmetry boundary here, I think the limiter can break it. So I suspect we need to do it after the limiter has been applied. I think it's OK to violate the limiter in this case.

Reply:
Actually, this is OK. The second and fourth order limiters will always respect a slope of zero and ensure the final slope is zero.